1385 - Collect statistics by default in ConvertToDelta & Update SQL API

amirmor1 opened the pull request:
## Description

Resolves https://github.com/delta-io/delta/issues/1385

This PR gives the option to collect column statistics per file when doing ConvertToDelta operation. The current behaviour should stay the same (meaning that statistics should not be collected on convert).


## How was this patch tested?

Current unit tests + added a couple of unit tests to cover this change

## Does this PR introduce _any_ user-facing changes?

Yes. Previously ConvertToDelta command was without the ability to collect statistics. Now, the convert function is overloaded with additional boolean parameter (with false as default) to enable collect statistics.
When this parameter is set to true, after the creation of the table, delta will compute all the files statistics (using legacy functionality and create another COMPUTE STATS commit to the log)

scottsand-db commented on core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:
```diff
@@ -79,7 +79,7 @@ statement
         (LIMIT limit=INTEGER_VALUE)?                                    #describeDeltaHistory
     | CONVERT TO DELTA table=qualifiedName
-        (PARTITIONED BY '(' colTypeList ')')?                           #convert
+        (PARTITIONED BY '(' colTypeList ')')? (STATS)?                          #convert
```
nit: fix indentation of `#convert` to align with other comments

scottsand-db replied:
I think we will need to get some input/feedback on this SQL API.

scottsand-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -577,8 +577,16 @@ object DeltaTable {
       spark: SparkSession,
       identifier: String,
       partitionSchema: StructType): DeltaTable = {
+    convertToDelta(spark, identifier, partitionSchema, false)
```
nit: named boolean params please

scottsand-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -604,8 +612,17 @@ object DeltaTable {
       spark: SparkSession,
       identifier: String,
       partitionSchema: String): DeltaTable = {
+    convertToDelta(spark, identifier, partitionSchema, false)
```
ditto: named boolean params

scottsand-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -628,8 +645,15 @@ object DeltaTable {
   def convertToDelta(
       spark: SparkSession,
       identifier: String): DeltaTable = {
+    convertToDelta(spark, identifier, false)
```
ditto

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -19,25 +19,23 @@ package org.apache.spark.sql.delta.commands
 // scalastyle:off import.ordering.noEmptyLine
 import java.io.Closeable
 import java.util.Locale
-
```
nit: add back empty line between different import groups

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -19,25 +19,23 @@ package org.apache.spark.sql.delta.commands
 import java.util.Locale
-
 import scala.collection.JavaConverters._
-
```
here too

scottsand-db replied:
@amirmor1 are you going to add back these blank spaces?

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -19,25 +19,23 @@ package org.apache.spark.sql.delta.commands
 import org.apache.spark.sql.delta.util._
 import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.fs.{FileStatus, FileSystem, Path}
-
```
here too

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -70,7 +68,8 @@ import org.apache.spark.util.SerializableConfiguration
     partitionSchema: Option[StructType],
-    deltaPath: Option[String]) extends LeafRunnableCommand with DeltaCommand {
+    deltaPath: Option[String],
+    collectStats: Boolean) extends LeafRunnableCommand with DeltaCommand {
```
add `collectStats` scaladoc above

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -394,8 +394,9 @@ abstract class ConvertToDeltaCommandBase(
-    deltaPath: Option[String])
-  extends ConvertToDeltaCommandBase(tableIdentifier, partitionSchema, deltaPath)
+    deltaPath: Option[String],
+    collectStats: Boolean = false)
```
why is the default false **here** instead of in `ConvertToDeltaCommandBase`?

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaScalaSuite.scala:
```diff
@@ -21,18 +21,21 @@ import org.apache.spark.sql.types.StructType
       )
     }
   }
+
```
nit: delete whitespace noise

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSuiteBase.scala:
```diff
@@ -98,6 +102,47 @@ trait ConvertToDeltaSuiteBase extends ConvertToDeltaSuiteBaseCommons
+    withTempDir { dir =>
+      val tempDir = dir.getCanonicalPath
+      writeFiles(tempDir, simpleDF)
+        convertToDelta(s"parquet.`$tempDir`", collectStats = true)
+        val deltaLog = DeltaLog.forTable(spark, tempDir)
+        val history = io.delta.tables.DeltaTable.forPath(tempDir)
+          .history()
```
fix this indentation

scottsand-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -577,8 +577,16 @@ object DeltaTable {
+    convertToDelta(spark, identifier, partitionSchema, false)
+  }
+
+  def convertToDelta(
```
can you also please update https://github.com/delta-io/delta/blob/master/python/delta/tables.py ? we need to keep parity between DeltaTable.scala (spark API) and our PySpark API.

scottsand-db reviewed (changes requested):
Looks great! Thanks for doing this!

Left some nits about styling and named boolean params.

Tests look good.

Most importantly, we need to update our PySpark API inside of `/pythong/delta/tables.py`, too.

scottsand-db commented:
@amirmor1 what if we make the default implementation collect stats, and we have to opt out from doing so. This is a one-time per table operation, and I'd expect that users would prefer the default option of having faster reads henceforth with dataskipping and stats, instead of a slightly faster CONVERT operation.

Thus, the SQL API could look like `CONVERT TO DELTA table_name [ NO STATISTICS ] [ PARTITIONED BY clause ]`

amirmor1 commented:
@scottsand-db Thanks for the review and the comments.
I've fixed the PR with your requests.

scottsand-db reviewed (changes requested):
Hi @amirmor1 - sorry to propose a bit of a change, but I'm going to ask you to do the following for this PR

- can you change this PR to **just** setting statistics-collection to true by default AND including the `NO STATISTICS` SQL API? we will merge this PR first.
- then, can you refactor the `DeltaTable` API changes into a separate PR? we will merge this PR second.

also, the SQL API change is `NO STATISTICS`. not `NO_STATISTICS`.

amirmor1 commented:
@scottsand-db just want to make sure that I understand. You want the collectStats to be true by default but without any mention in the `DeltaTable` API? If yes -> So on the SQL API the `NO STATISTICS` has no effect, right? (at least until the second PR).
If yes -> it also means that a multiple unit tests which are not directly related to this PR will need to change since they expect to have a single commit on create while now we'll have two commits (I was able to avoid changing the results with `NO STATISTICS` during last commit).

scottsand-db commented:
> just want to make sure that I understand. You want the collectStats to be true by default but without any mention in the DeltaTable API? If yes -> So on the SQL API the NO STATISTICS has no effect, right? (at least until the second PR).

- I want collectStats to be true by default.
- In the SQL API, when the user includes NO STATISTICS like `CONVERT TO DELTA table_name [ NO STATISTICS ] [ PARTITIONED BY clause ]` I want it to have an **effect**. I want statistics to **not** be collected.

In the second PR, you will add the DeltaTable API.

scottsand-db commented:
>  it also means that a multiple unit tests which are not directly related to this PR will need to change since they expect to have a single commit on convert while now we'll have two commits

You can change these tests to use the `NO STATISTICS` SQL API. 

And then add **new** tests which do not have `NO STATISTICS`.

Make sense? Thanks for the great clarifying questions!

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -356,6 +358,7 @@ abstract class ConvertToDeltaCommandBase(
         spark.sessionState.catalog
       )
     }
+    if (collectStats) { StatisticsCollection.recompute(spark, txn.deltaLog) }
```
hm. so this would be done in the **next** commit? what if there's a commit in-between these two?

amirmor1 replied:
Good question. 

So files that were added after the first commit of the convert, should already have statistics in case the default configuration wasn't changed (which makes me think that I should also check if the configuration is set to collect statistics before actually collecting them. This raises another question - what if collectStats is true but config is set to false? Should we raise an exception here or simply avoid collecting stats during convert?)

So the files on first commit does not have stats, files on second commit (the one between the convert and compute-stats) have statistics by default, now `StatisticsCollection.recompute` should take all the files from the log and collect their statistics (I think that the files from the second commit will be re-computed).

Eventually all the files should have statistics after the third commit.

I tested it locally with

1. writing parquet files
2. convert to delta without statistics
3. write delta files
4. recompute statistics
5. all files had statistics

I can add this as an additional unit test.

Are you asking because maybe the `txn.deltaLog` is not updated with the latest files from commit 2? I admit I'm not yet familiar with the code base so I'm not sure if that is possible and how, but maybe `StatisticsCollection.recompute(spark, txn.deltaLog.update().deltaLog)` should make sure no files will be missing?

scottsand-db replied:
Sorry for the delay in response.

> what if collectStats is true but config is set to false? Should we raise an exception here or simply avoid collecting stats during convert?

Let's throw an exception. This is a weird state, we don't want to confuse the user.

> Are you asking because maybe the txn.deltaLog is not updated with the latest files from commit 2?

I'm asking because of the edge case of 
```
commit 0 -> convert parquet to delta
commit 1 -> remove a file
commit 2 -> compute stats finally finishes. COLLISION.
```

> Are you asking about the possibility of a conflict on the next commit (the stats commit) which could eventually throw an exception? So in that case we'll have a converted delta table but without statistics even though collectStats was set to true? If so -> Should we raise another, more specific exception do describe the situation? Or maybe call recompute again?

No don't throw an exception. And don't call recompute again. I think this case is very, very rare.

But ...... I do want to think about if we can do this in only 1 commit instead of 2.

scottsand-db replied:
I've figured out a simple way to write the stats in 1 commit instead of 2.

Let's add this to the `ConvertToDeltaCommand` object:
```scala
def computeStats(
      deltaLog: DeltaLog,
      snapshot: Snapshot,
      addFiles: Seq[AddFile]): Iterator[AddFile] = {
    val filesWithStats = deltaLog.createDataFrame(snapshot, addFiles)
      .groupBy(input_file_name()).agg(to_json(snapshot.statsCollector))

    val pathToAddFileMap = generateCandidateFileMap(deltaLog.dataPath, addFiles)
    filesWithStats.collect().iterator.map { row =>
      val addFile = getTouchedFile(deltaLog.dataPath, row.getString(0), pathToAddFileMap)
      addFile.copy(stats = row.getString(1))
    }
  }
```

and then we can call it from inside of `ConvertToDeltaCommandBase::createDeltaActions`.

i.e. if if we should collect stats, return the method above, else inside of `createDeltaActions` we just return the current existing `adds.toIterator`.

make sense?

scottsand-db commented:
Can you update the PR title to be `Collect statistics by default in ConvertToDelta & Update SQL API`

amirmor1 commented:
Hi @scottsand-db.
Thanks for the comments and suggestion to make this on a single commit.
I've added the changes and also added a log message when collectStats is enabled but this option is disabled on config. Figured it requires some "explenation" why no statistics will be in such case.
Let me know if it makes sense.

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSuiteBase.scala:
```diff
@@ -16,36 +16,39 @@
 
-  protected def convertToDelta(identifier: String, partitionSchema: Option[String] = None): Unit
+  protected def convertToDelta(identifier: String, partitionSchema: Option[String] = None,
+      collectStats: Boolean = false): Unit
```
since we want the default & most common usage to be `collectStats=true` should't this default be `true` ?

do any tests break?

i'd like to test the most common usage ideally.

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -360,6 +375,21 @@ abstract class ConvertToDeltaCommandBase(
+    }
+  }
+
   protected def getContext: Map[String, String] = {
```
you also need to update this below on line 397

```scala
/** Get the operation to store in the commit message. */
  protected def getOperation(
      numFilesConverted: Long,
      convertProperties: ConvertTarget): DeltaOperations.Operation = {
    DeltaOperations.Convert(
      numFilesConverted,
      partitionSchema.map(_.fieldNames.toSeq).getOrElse(Nil),
      statsEnabled && collectStats <<<--------------!!!!!
      convertProperties.catalogTable.map(t => t.identifier.toString))
  }
```

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -394,8 +424,9 @@ abstract class ConvertToDeltaCommandBase(
-    deltaPath: Option[String])
-  extends ConvertToDeltaCommandBase(tableIdentifier, partitionSchema, deltaPath)
+    deltaPath: Option[String],
+    collectStats: Boolean)
```
```scala
collectStats: Boolean,
deltaPath: Option[String]
```

scottsand-db reviewed (changes requested):
Some more minor comments. Looking good and thanks for working on this!

amirmor1 commented:
@scottsand-db Thanks again for the help!
I've pushed the requested changes

scottsand-db commented on core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:
```diff
@@ -166,7 +166,7 @@ nonReserved
     | RESTORE | AS | OF
     | ZORDER | LEFT_PAREN | RIGHT_PAREN
-    | SHOW | COLUMNS | IN | FROM
+    | SHOW | COLUMNS | IN | FROM | NO | STATISTICS
```
@jackierwzhang do you think this is okay here?

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -266,16 +269,30 @@ abstract class ConvertToDeltaCommandBase(
+      fs: FileSystem,
+      collectStats: Boolean): Iterator[AddFile] = {
+    val initialSnapshot = new InitialSnapshot(txn.deltaLog.dataPath, txn.deltaLog, txn.metadata)
+    val collectStatsDeltaConfiguration = sparkSession.sessionState.conf
```
you don't need `sparkSession`. you can just do `conf.getConf`

also let's name this variable `statsEnabled`

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -266,16 +269,30 @@ abstract class ConvertToDeltaCommandBase(
-      adds.toIterator
+      if (shouldCollectStats) {
+        computeStats(txn.deltaLog, initialSnapshot, adds)
+      }
```
let's use the style
```scala
if () {
...
} else if () {
...
}
```

instead of 
```scala
if () {
...
}
else if () {
...
}
```
```

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaScalaSuite.scala:
```diff
@@ -21,18 +21,16 @@ import org.apache.spark.sql.types.StructType
       io.delta.tables.DeltaTable.convertToDelta(
         spark,
         identifier,
-        StructType.fromDDL(partitionSchema.get)
```
revert these whitespace changes. it's just noise.

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSuiteBase.scala:
```diff
@@ -16,36 +16,39 @@
+import org.apache.spark.sql.{AnalysisException, DataFrame, QueryTest, Row}
 import org.apache.spark.util.Utils
 
+import java.io.{File, FileNotFoundException}
```
java imports come before `org.apache.hadoop` imports.

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaScalaSuite.scala:
```diff
@@ -21,18 +21,16 @@ import org.apache.spark.sql.types.StructType
   override protected def convertToDelta(
       identifier: String,
-      partitionSchema: Option[String] = None): Unit = {
+      partitionSchema: Option[String] = None, collectStats: Boolean = false): Unit = {
```
didn't I comment last time that we wanted this to be `true` by default?

the new default is `true` ... so we want most of our tests to explicitly test for that.

scottsand-db replied:
thus your new test should be one where it is explicitly set to false.

how many tests break when you set the default to true?

amirmor1 replied:
Sorry, missed that. There are no tests that break because of this change in this class.

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/DeltaColumnMappingTestUtils.scala:
```diff
@@ -350,7 +350,7 @@ trait DeltaColumnMappingTestUtilsBase extends SharedSparkSession {
    */
   protected def convertToDelta(tableOrPath: String): Unit = {
-    sql(s"CONVERT TO DELTA $tableOrPath")
+    sql(s"CONVERT TO DELTA $tableOrPath NO STATISTICS")
```
why this change?

amirmor1 replied:
Since eventually we convert to delta with stats on a single commit, there is no need for that to make the tests pass. Removed.

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala:
```diff
@@ -247,7 +247,7 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
     testTableCreation { dir =>
       spark.range(10).write.mode("append").parquet(dir)
-      sql(s"CONVERT TO DELTA parquet.`$dir`")
+      sql(s"CONVERT TO DELTA parquet.`$dir` NO STATISTICS")
```
same here. why this change?

amirmor1 replied:
Since eventually we convert to delta with stats on a single commit, there is no need for that to make the tests pass. Removed.

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/DescribeDeltaHistorySuite.scala:
```diff
@@ -924,7 +924,9 @@ trait DescribeDeltaHistorySuiteBase
         )
-        val operationMetrics = getOperationMetrics(deltaTable.history(1))
+        // TODO: Remove filter once Delta API is ready with collectStats in Convert
+        val operationMetrics = getOperationMetrics(deltaTable.history()
```
why this change. what is missing/failing without it?

amirmor1 replied:
Since eventually we convert to delta with stats on a single commit, there is no need for that to make the tests pass. Removed.

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSuiteBase.scala:
```diff
@@ -18,20 +18,19 @@ package org.apache.spark.sql.delta
 import org.apache.spark.sql.execution.streaming.MemoryStream
 import org.apache.spark.sql.functions._
-import org.apache.spark.sql.test.{SharedSparkSession, SQLTestUtils}
+import org.apache.spark.sql.test.{SQLTestUtils, SharedSparkSession}
```
revert this

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSuiteBase.scala:
```diff
@@ -98,6 +101,44 @@ trait ConvertToDeltaSuiteBase extends ConvertToDeltaSuiteBaseCommons
     }
   }
 
+  test("convert with collectStats true") {
```
so these tests are fine; but now that the default is `true` (and you've updated the test suites accordingly to reflect that) we need tests for when it is explicitly `false`.

amirmor1 replied:
I wonder how to achieve that in our `ConvertToDeltaScalaSuite`.

Both `ConvertToDeltaScalaSuite` and `ConvertToDeltaSQLSuite` inherite from `ConvertToDeltaSuiteBase`, but since we added this API to the SQL function but still not to the scala `DeltaTable` (its in the second PR), how can we do that?

`HiveConvertToDeltaSuiteBase` does tests collectStats as false

scottsand-db reviewed (changes requested):
Getting close! Asked for some trivial changes, a minor refactor (move a method to an object function) and some more tests.

Thanks for this! Looking forward to merging this soon

amirmor1 commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -363,6 +377,21 @@ abstract class ConvertToDeltaCommandBase(
     Seq.empty[Row]
   }
 
+  private def computeStats(
```
In that case I think I'll need to delete the `protected` from the following functions: `getTouchedFile`, `generateCandidateFileMap`

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/DeltaCommand.scala:
```diff
@@ -91,7 +91,7 @@ trait DeltaCommand extends DeltaLogging {
    * each file contains a UUID.
    */
-  protected def generateCandidateFileMap(
+   def generateCandidateFileMap(
```
nit I think this needs to be indented back 1 space right?

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/DeltaCommand.scala:
```diff
@@ -154,7 +154,7 @@ trait DeltaCommand extends DeltaLogging {
    * @param nameToAddFileMap Map generated through `generateCandidateFileMap()`
    */
-  protected def getTouchedFile(
+   def getTouchedFile(
```
here too: indent back 1 space

scottsand-db reviewed (changes requested):
Some more minor nits. After this, LGTM! Thanks so much!

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:
```diff
@@ -65,11 +67,13 @@ import org.apache.spark.util.SerializableConfiguration
 abstract class ConvertToDeltaCommandBase(
     tableIdentifier: TableIdentifier,
     partitionSchema: Option[StructType],
+    collectStats: Boolean = true,
```
I have no idea how this is compiling, params with default values must go at the end of the param list.

anyways, please remove this default.

scottsand-db replied:
also, please resolve conversations once you have addressed the comment :) this will make viewing the PR much easier.

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSQLSuite.scala:
```diff
@@ -16,20 +16,47 @@
   override protected def convertToDelta(
       identifier: String,
-      partitionSchema: Option[String] = None): Unit = {
+      partitionSchema: Option[String] = None, collectStats: Boolean = true): Unit = {
```
are all the existing tests now run with `collectStats=true`?

amirmor1 replied:
Yes, except from one under SQLSuite that checks false

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSQLSuite.scala:
```diff
@@ -16,20 +16,47 @@
+      withSQLConf(DeltaSQLConf.DELTA_COLLECT_STATS.key -> "true") {
+
+        val tempDir = dir.getCanonicalPath
+        writeFiles(tempDir, simpleDF)
```
How many files does this generate? Can we have more than one file in the test if not already

amirmor1 replied:
It creates 2 files

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSQLSuite.scala:
```diff
@@ -16,20 +16,47 @@
+        val statsDf = deltaLog.unsafeVolatileSnapshot.allFiles
+          .select(from_json(col("stats"), deltaLog.unsafeVolatileSnapshot.statsSchema)
+            .as("stats")).select("stats.*")
+        assert(statsDf.filter(col("numRecords").isNotNull).count == 0)
```
we could add another assert that checks the sum of `numRecords` is equal to the number of records in `simpleDF`

amirmor1 replied:
Done

vkorukanti commented on core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:
```diff
@@ -230,7 +230,7 @@ class DeltaSqlAstBuilder extends DeltaSqlBaseBaseVisitor[AnyRef] {
       visitTableIdentifier(ctx.table),
       Option(ctx.colTypeList).map(colTypeList => StructType(visitColTypeList(colTypeList))),
-      None)
+      ctx.STATISTICS() == null, None)
```
Sorry this is not clear, does this returns null when `NO STATISTICS` is present?

amirmor1 replied:
When `NO STATISTICS` is present, this returns false. Otherwise true.

vkorukanti reviewed (commented):
lgtm

scottsand-db commented:
@amirmor1 - hope all is well! Just following up on this PR. It's close to being merged and we would LOVE to have this in delta 2.2!

amirmor1 commented:
Hi @scottsand-db, I'm all good, thanks. Sorry for the late reply but I'm currently on vacation until next Tuesday.

scottsand-db commented:
@amirmor1 - no worries!! Enjoy the vacation! 🍹

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSuiteBase.scala:
```diff
@@ -98,6 +101,47 @@ trait ConvertToDeltaSuiteBase extends ConvertToDeltaSuiteBaseCommons
+    }
+  }
+
+  test("convert with collectStats true but config set to false") {
```
we expect this to write stats, correct? can you add a comment explaining that we expect `collectStats=true` to override the default conf value.

scottsand-db replied:
oh I see. `DELTA_COLLECT_STATS` is whether stats collection is enabled in general.

this makes sense!

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/ConvertToDeltaSuiteBase.scala:
```diff
@@ -98,6 +101,47 @@ trait ConvertToDeltaSuiteBase extends ConvertToDeltaSuiteBaseCommons
+      }
+    }
+  }
+
```
is there no test for when `collectStats=false` ?

scottsand-db commented on core/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala:
```diff
@@ -191,7 +191,7 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
 
           val deltaLog = DeltaLog.forTable(spark, dir)
-          assert(deltaLog.snapshot.version === 0, "did not create a Delta table")
+          assert(deltaLog.snapshot.version == 0, "did not create a Delta table")
```
revert this change, `===` is a helpful debug tool.

scottsand-db reviewed (changes requested):
Trivial minor comments. After these are resolved - LGTM!

scottsand-db reviewed (changes requested):
Can you please fix the failing tests? https://github.com/delta-io/delta/actions/runs/3468626879/jobs/5794653322

You just need to all `collectStats = true` to the `ConvertToDeltaCommand` calls inside of `delta-iceberg/src/test/scala/org/apache/spark/sql/delta/ConvertIcebergToDeltaSuite.scala`.

```
[error] /home/runner/work/delta/delta/delta-iceberg/src/test/scala/org/apache/spark/sql/delta/ConvertIcebergToDeltaSuite.scala:188:30: not enough arguments for method apply: (tableIdentifier: org.apache.spark.sql.catalyst.TableIdentifier, partitionSchema: Option[org.apache.spark.sql.types.StructType], collectStats: Boolean, deltaPath: Option[String]): org.apache.spark.sql.delta.commands.ConvertToDeltaCommand in object ConvertToDeltaCommand.
[error] Unspecified value parameter deltaPath.
[error]         ConvertToDeltaCommand(
[error]                              ^
[error] /home/runner/work/delta/delta/delta-iceberg/src/test/scala/org/apache/spark/sql/delta/ConvertIcebergToDeltaSuite.scala:744:32: not enough arguments for method apply: (tableIdentifier: org.apache.spark.sql.catalyst.TableIdentifier, partitionSchema: Option[org.apache.spark.sql.types.StructType], collectStats: Boolean, deltaPath: Option[String]): org.apache.spark.sql.delta.commands.ConvertToDeltaCommand in object ConvertToDeltaCommand.
[error] Unspecified value parameter deltaPath.
[error]           ConvertToDeltaCommand(
[error]                                ^
[error] /home/runner/work/delta/delta/delta-iceberg/src/test/scala/org/apache/spark/sql/delta/ConvertIcebergToDeltaSuite.scala:876:26: not enough arguments for method apply: (tableIdentifier: org.apache.spark.sql.catalyst.TableIdentifier, partitionSchema: Option[org.apache.spark.sql.types.StructType], collectStats: Boolean, deltaPath: Option[String]): org.apache.spark.sql.delta.commands.ConvertToDeltaCommand in object ConvertToDeltaCommand.
[error] Unspecified value parameter deltaPath.
[error]     ConvertToDeltaCommand(
[error]                          ^
[error] /home/runner/work/delta/delta/delta-iceberg/src/test/scala/org/apache/spark/sql/delta/ConvertIcebergToDeltaSuite.scala:1226:26: not enough arguments for method apply: (tableIdentifier: org.apache.spark.sql.catalyst.TableIdentifier, partitionSchema: Option[org.apache.spark.sql.types.StructType], collectStats: Boolean, deltaPath: Option[String]): org.apache.spark.sql.delta.commands.ConvertToDeltaCommand in object ConvertToDeltaCommand.
[error] Unspecified value parameter deltaPath.
[error]     ConvertToDeltaCommand(
[error]  
```

scottsand-db commented:
Hi @amirmor1 could you please provide me with a public email address? I can't merge this with your private email address '49829354+amirmor1@users.noreply.github.com'. This is against Linux Foundation guidelines.

amirmor1 commented:
@scottsand-db zhify1@gmail.com
